### PR TITLE
Fix flush() in clean() and refresh()

### PR DESCRIPTION
**Pull request by lrq3000**
Continuation of #327 by @deeenes.

We now use `status_printer()` in `clean()` and `refresh()`, so not only this automatically does `flush()`, but it also saves the last print length, which will provide a more consistent display after executing `clean()` or `refresh()` and then displaying the bar again (I guess doing `clean()` and `update()` in this order might produce weird displays sometimes, this should be fixed by this PR).

Just one test for `write()` had to be adapted, I checked myself with old example scripts using `write()` in various sceniarios, everything looks OK.

TODO:
- Nothing.

### CODE REVIEW

File: tqdm/_tqdm.py [hunk 1]
```diff
@@ -1049,8 +1049,7 @@ def clear(self, nomove=False):
         if not nomove:
             self.moveto(self.pos)
         # clear up the bar (can't rely on sp(''))
-        self.fp.write('\r')
-        self.fp.write(' ' * (self.ncols if self.ncols else 10))
+        self.sp('')
```
Comment (casperdcl) [inline tqdm/_tqdm.py]: is L1052 sufficient despite the warning on L1051?

Comment (lrq3000) [inline tqdm/_tqdm.py]: Oh, I totally missed the comment! All the unit tests pass fine and my own test scripts too, so I don't know nor remember at all why sp couldn't be relied on. I'm going to use git blame and git pickaxe to trace back this comment and see whether I can get more info.

**Comment by lrq3000** [comment]
Just a quick update, please don't merge this one yet. I've investigated a bit and tried to use `sp()` everywhere instead of direct writing, but 2 test are now failing, the output being different, and it's not really obvious why. I guess the unreliability of `sp('')` is still true but I don't know why as it should work, I need to hunt down this bug some more!

**Comment by casperdcl** [comment]
I recall writing this code a long time ago and there was a good reason. I think. Which was why I asked. It's better if you take a look at it for a second opinion.

**Comment by LinxiFan** [comment]
Is this PR gonna be merged soon? I think it's the cause for #420 as well (in the latest 4.15 release).

**Comment by lrq3000** [comment]
As an alternative to this PR, we could just use `self.fp.flush()`, this would basically have the same goal (without the weird side effects of `self.sp('')`.

**Comment by MPagel** [comment]
I tried self.fp.flush() first, but that didn't seem to work in a multithreading environment (eventlet) where I was catching all output to stdout/err as per documentation.  The following seems to work:     

```
def flush(self):
        fp_flush = getattr(self.file, 'flush', lambda: None)
        fp_flush()
```

See also #439 #329 ab1bd789f143b79da7b7c18ba942269afcd7df7d

**Comment by casperdcl** [comment]
could you try the `devel` branch to verify it fixes this?

**Pull request closed without merging**